server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -646,8 +674,41 @@ public void moveShards() {
             // Iterate over the started shards interleaving between nodes, and check if they can remain. In the presence of throttling
             // shard movements, the goal of this iteration order is to achieve a fairer movement of shards from the nodes that are
             // offloading the shards.
+
+            // Trying to eliminate target nodes so that we donot unnecessarily iterate over source nodes
+            // when no target is eligible
+            for (ModelNode currentNode : sorter.modelNodes) {
+                RoutingNode target = currentNode.getRoutingNode();
+                Decision nodeLevelAllocationDecision = allocation.deciders().canAllocateAnyShardToNode(target, allocation);
+                if (nodeLevelAllocationDecision.type() != Decision.Type.YES) {
+                    inEligibleTargetNode.add(currentNode.getRoutingNode());
+                }
+            }
             for (Iterator<ShardRouting> it = allocation.routingNodes().nodeInterleavedShardIterator(); it.hasNext(); ) {
                 ShardRouting shardRouting = it.next();
+                shardsProcessed++;
```

Comment:
Should this be updated only after the early termination check for the node?

Reply:
Makes sense. Updated!